x-pack/plugins/fleet/server/services/actions/actions.ts

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { v4 as uuidV4 } from 'uuid';
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import { fromKueryExpression, toElasticsearchQuery } from '@kbn/es-query';
+import type { estypes } from '@elastic/elasticsearch';
+
+import { ES_SEARCH_LIMIT } from '../../../common/constants';
+
+import { FleetActionsError } from '../../../common/errors';
+import { AGENT_ACTIONS_INDEX, AGENT_ACTIONS_RESULTS_INDEX } from '../../../common';
+import { auditLoggingService } from '../audit_logging';
+
+import {
+  validateFilterKueryNode,
+  allowedFleetActionsFields,
+  ALLOWED_FLEET_ACTIONS_FIELD_TYPES,
+} from './utils';
+
+import type { FleetActionRequest, FleetActionResult, BulkCreateResponse } from './types';
+
+const queryOptions = Object.freeze({
+  ignore: [404],
+});
+
+export const createAction = async (
+  esClient: ElasticsearchClient,
+  action: FleetActionRequest
+): Promise<FleetActionRequest> => {
+  try {
+    const body = {
+      ...action,
+      action_id: action.action_id || uuidV4(),
+      '@timestamp': action['@timestamp'] || new Date().toISOString(),
+    };
+    await esClient.create(
+      {
+        index: AGENT_ACTIONS_INDEX,
+        // doc id is same as action_id
+        id: body.action_id,
+        body,
+        refresh: 'wait_for',
+      },
+      {
+        meta: true,
+      }
+    );
+
+    auditLoggingService.writeCustomAuditLog({
+      message: `User created Fleet action [id=${action.action_id}, user_id=${action.user_id}, input_type=${action.input_type}]`,
+    });
+
+    return body;
+  } catch (createActionError) {
+    throw new FleetActionsError(
+      `Error creating action: ${createActionError.message}`,
+      createActionError
+    );
+  }
+};
+
+const getLoggingInfo = ({
+  id,
+  actions,
+}: {
+  id: string;
+  actions: FleetActionRequest[];
+}): {
+  input_type: string;
+  user_id: string;
+} => {
+  const action = actions.filter((item) => item.action_id === id)[0];
```

Comment:
nit: nullcheck could be added here, if the action is not found

Reply:
Should we not log anything if there's no action found?

Reply:
It's probably not going to happen, as the id is coming from the bulk create response. Maybe just defaulting to `""` in logging info is enough.

Reply:
done 67e7f3ef074e0ec4f6f75179720e6856daf4db1f